### PR TITLE
Rewrite OWNERS docs

### PR DESCRIPTION
**Pull request by spiffxp**
Move from design doc to state-of-the-world doc

The OWNERS feature has been implemented for a while now. Let's
document how it actually works

I originally started this with the intent of fleshing out measurable
critera for adding/removing OWNERS, but as I dug into it, I realized I
need more time to explore useful metrics.  Let's find out if this is
enough of a rewrite to call incremental progress

ref: https://github.com/kubernetes/test-infra/issues/1966

### CODE REVIEW

**Comment by spiffxp** [comment]
/cc @fejta @grodrigues3 @thockin @pwittrock @apelisse @cjwagner 

I'm not entirely sure whose approval I need to move this forward, I've talked to some of you, and others I'm gleaning from commits.  PTAL.

File: contributors/devel/owners.md [hunk 1]
```diff
@@ -1,100 +1,228 @@
...
+We should strive to:
+
+- grow the number of OWNERS files
+- add new people to OWNERS files
+- keep people in OWNERS files who are actively contributing to or reviewing the code they own
+- remove inactive people from OWNERS files
+
+Bad examples of OWNERS usage:
+
+- directories that lack OWNERS files, resulting in too many hitting root OWNERS
```
Comment (spiffxp) [inline contributors/devel/owners.md]: ref: https://github.com/kubernetes/kubernetes/issues/44420

Comment (spiffxp) [inline contributors/devel/owners.md]: ref: https://github.com/kubernetes/kubernetes/issues/35425

File: contributors/devel/owners.md [hunk 2]
```diff
@@ -1,100 +1,228 @@
...
+
+
+## Updating OWNERS
+
+OWNERS files should be regularly maintained.
+
+We should strive to:
+
+- grow the number of OWNERS files
+- add new people to OWNERS files
```
Comment (spiffxp) [inline contributors/devel/owners.md]: ref: https://github.com/kubernetes/kubernetes/pull/46125

Comment (spiffxp) [inline contributors/devel/owners.md]: ref: https://github.com/kubernetes/kubernetes/pull/49166

File: contributors/devel/owners.md [hunk 3]
```diff
@@ -1,100 +1,228 @@
...
+- keep people in OWNERS files who are actively contributing to or reviewing the code they own
+- remove inactive people from OWNERS files
```
Comment (spiffxp) [inline contributors/devel/owners.md]: ref: https://github.com/kubernetes/kubernetes/pull/47280

File: contributors/devel/owners.md [hunk 4]
```diff
@@ -1,100 +1,228 @@
...
-```
-@k8s-bot approvers
-```
+OWNERS files support the following keys:
 
-into a comment.
+- `approvers`: a list of github usernames or aliases that can `/approve` a PR
+- [DEPRECATED] `assignees`: do not use, equivalent to `approvers`
+  ([kubernetes/test-infra#3851](https://github.com/kubernetes/test-infra/issues/3851)) 
+- `labels`: a list of github labels to automatically apply to a PR
```
Comment (spiffxp) [inline contributors/devel/owners.md]: ref: https://github.com/kubernetes/kubernetes/pull/49160

File: contributors/devel/owners.md (hunk 3 above)
Comment (spiffxp) [inline contributors/devel/owners.md]: ref: https://github.com/kubernetes/kubernetes/pull/48886

File: contributors/devel/owners.md [hunk 5]
```diff
@@ -1,100 +1,228 @@
...
+    issue provides value
+  - Protest is being expressed in the form of overuse of `/approve no-issue`
+  - Instead, suggest to the PR **author** that they edit the PR description to include a linked
+    issue
+  - This is a sign that we need to actually deliver value with linked issues, or be able to define
+    what a "trivial" PR is in a machine-enforceable way to be able to automatically waive the linked
+issue requirement
+- Technically, anyone who is a member of the kubernetes GitHub organization can drive-by review a
+  PR.
+  - This goes against the idea of specifying **reviwers** in the first place, to ensure that the
```
Comment (cblecker) [inline contributors/devel/owners.md]: The wording of this and it being listed in the "Quirks" section makes it sound like it's a bad thing.. however, reviewing PRs is one of the ways contributors can/are required to ascend the contributor ladder.

File: contributors/devel/owners.md [hunk 6]
```diff
@@ -1,100 +1,228 @@
...
+  - The PR is merged
+
+## Quirks of the Process
+
+There are a number of behaviors we've observed that while _possible_ are discouraged, as they go
+against the intent of this review process.  Some of these could be prevented in the future, but this
+is the state of today.
+
+- An **approver**'s `/lgtm` is simultaneously interpreted as an `/approve`.
+  - This goes against the idea of having two sets of eyes on a PR, and can be confusing to an
```
Comment (cblecker) [inline contributors/devel/owners.md]: Another quirk would be when the approver selected by mungebot is also one of the two reviewers selected. I don't know if there is a better way to avoid this.

File: contributors/devel/owners.md (hunk 5 above)
Comment (spiffxp) [inline contributors/devel/owners.md]: I struggle with this... how else do you qualify as a reviewer other than reviewing?  We have to allow this

On the other hand, getting feedback from someone who doesn't understand the code can cause a PR author to waste their time addressing feedback that won't actually move the PR forward.  How do we make sure they know who needs to be paid attention to?

Comment (cjwagner) [inline contributors/devel/owners.md]: s/reviwers/reviewers/

File: contributors/devel/owners.md [hunk 7]
```diff
@@ -1,100 +1,228 @@
...
+  - Instead, suggest to the PR **author** that they edit the PR description to include a linked
+    issue
+  - This is a sign that we need to actually deliver value with linked issues, or be able to define
+    what a "trivial" PR is in a machine-enforceable way to be able to automatically waive the linked
+issue requirement
+- Technically, anyone who is a member of the kubernetes GitHub organization can drive-by review a
+  PR.
+  - This goes against the idea of specifying **reviwers** in the first place, to ensure that the
+    **author** is getting actionable feedback from people knowledgeable with the code.
+  - This is a sign that our OWNERS files are too small; growing the `reviewers` list would whelp
```
Comment (marun) [inline contributors/devel/owners.md]: s/whelp/help/

File: contributors/devel/owners.md [hunk 8]
```diff
@@ -1,100 +1,228 @@
...
+  gitrepos](https://github.com/kubernetes/test-infra/blob/master/mungegithub/features/repo-updates.go)
+  - responsible for parsing OWNERS files
+  - if its `use-reviewers` flag is set to false, **approvers** will also be **reviewers**
+  - it its `enable-md-yaml` flag is set, `.md` files will also be parsed to see if they have
+    embedded OWNERS content (this is only used by
+[kubernetes.github.io](https://github.com/kubernetes/kubernetes.github.io/))
+  - used by other mungers to get the set of **reviewers** or **approvers** for a given path
+- [feature:
+  aliases](https://github.com/kubernetes/test-infra/blob/master/mungegithub/features/aliases.go)
+  - repsonsible for parsing the OWNERS_ALIASES file
```
Comment (marun) [inline contributors/devel/owners.md]: s/repsonsible/responsible/

File: contributors/devel/owners.md [hunk 9]
```diff
@@ -1,100 +1,228 @@
...
+### [`mungegithub`](https://github.com/kubernetes/test-infra/tree/master/mungegithub)
+
+Mungegithub polls GitHub, and "munges" things it finds, including issues and pull requests. It is
+stateful, in that restarting it means it loses track of which things it has munged at what time.
+
+- [feature:
+  gitrepos](https://github.com/kubernetes/test-infra/blob/master/mungegithub/features/repo-updates.go)
+  - responsible for parsing OWNERS files
+  - if its `use-reviewers` flag is set to false, **approvers** will also be **reviewers**
+  - it its `enable-md-yaml` flag is set, `.md` files will also be parsed to see if they have
```
Comment (cjwagner) [inline contributors/devel/owners.md]: s/it/if/

File: contributors/devel/owners.md [hunk 10]
```diff
@@ -1,100 +1,228 @@
...
+  approval-handler](https://github.com/kubernetes/test-infra/blob/master/mungegithub/mungers/approval-handler.go)
+  - responsible for adding the  `approved` label once an **approver** for each of the required
+    OWNERS files has `/approve`'d
+  - responsible for commenting as required OWNERS files are satisfied
+  - responsible for removing outdated approval status comments
+  - [full description of the
+    algorithm](https://github.com/kubernetes/test-infra/blob/6f5df70c29528db89d07106a8156411068518cbc/mungegithub/mungers/approval-handler.go#L99-L111)
+- [munger:
+  submit-queue](https://github.com/kubernetes/test-infra/blob/master/mungegithub/mungers/submit-queue.go) 
+  - responsible for merging PR's<D-d><D-d>
```
Comment (cjwagner) [inline contributors/devel/owners.md]: Remove `<D-d><D-d>`? (Does this do something?)

File: contributors/devel/owners.md [hunk 11]
```diff
@@ -1,100 +1,228 @@
...
+
+OWNERS files should be regularly maintained.
+
+We should strive to:
+
+- grow the number of OWNERS files
+- add new people to OWNERS files
+- keep people in OWNERS files who are actively contributing to or reviewing the code they own
+- remove inactive people from OWNERS files
+
```
Comment (cjwagner) [inline contributors/devel/owners.md]: We should also try to get rid of all non-collaborators in OWNERS files. When the bots try to assign issues or PRs to those people they fail. The OWNERS files should really only contain members and outside collaborators.

File: contributors/devel/owners.md [hunk 12]
```diff
@@ -1,100 +1,228 @@
...
+  - This goes against the idea of specifying **reviwers** in the first place, to ensure that the
+    **author** is getting actionable feedback from people knowledgeable with the code.
+  - This is a sign that our OWNERS files are too small; growing the `reviewers` list would whelp
+    with this
+- **Reviewers**, and **approvers** are unresponsive
+  - This causes a lot of frustration for **authors** who often have little visibility into why their
+    PR is being ignored
+  - Many **reviewers** and **approvers** are so overloaded by GitHub notifications that @mention'ing
+    is unlikely to get a quick response
+  - If an **author** `/assign`'s a PR, it will show up on their **reviewers**' and **approvers**'
```
Comment (spiffxp) [inline contributors/devel/owners.md]: grammar is awkward, will rephrase

File: contributors/devel/owners.md (hunk 11 above)
Comment (spiffxp) [inline contributors/devel/owners.md]: agreed, ideally we could vet this when we add new entries to OWNERS files

**Comment by spiffxp** [comment]
I'd like to merge this tomorrow as something to iterate on unless there are any objections

File: contributors/devel/owners.md [hunk 13]
```diff
@@ -1,100 +1,232 @@
...
+## OWNERS spec
 
-After a PR is submitted, the automated kubernetes PR robot will append a message to the PR indicating the owners
-that are required for the PR to be submitted.
+The [mungegithub gitrepos
+feature](https://github.com/kubernetes/test-infra/blob/master/mungegithub/features/repo-updates.go)
+is the main consumer of OWNERS files.  If this page is out of date, look there.
 
-Subsequently, a user can also request the approval message from the robot by writing:
+Each directory that contains a unit of independent code or content also contains an OWNERS file.
```
Comment (thockin) [inline contributors/devel/owners.md]: "may also contain" ?

File: contributors/devel/owners.md [hunk 14]
```diff
@@ -1,100 +1,232 @@
...
+This file applies to everything within the directory, including the OWNERS file itself, sibling
+files, and child directories.
 
-```
-@k8s-bot approvers
-```
+OWNERS files support the following keys:
```
Comment (thockin) [inline contributors/devel/owners.md]: say that hey are YAML ?

File: contributors/devel/owners.md (hunk 6 above)
Comment (thockin) [inline contributors/devel/owners.md]: I dont see that as a problem

File: contributors/devel/owners.md [hunk 15]
```diff
@@ -1,100 +1,232 @@
...
+  - Instead, suggest to the PR **author** that they edit the PR description to include a linked
+    issue
+  - This is a sign that we need to actually deliver value with linked issues, or be able to define
+    what a "trivial" PR is in a machine-enforceable way to be able to automatically waive the linked
+issue requirement
+- Technically, anyone who is a member of the kubernetes GitHub organization can drive-by review a
+  PR.
+  - How else can you demonstrate your candidacy to be a reviewer other than by reviewing?  This seems 
+    like something we should continue to allow.  However...
+  - This goes against the idea of specifying **reviewers** in the first place, to ensure that the
```
Comment (thockin) [inline contributors/devel/owners.md]: I don't think it is a problem.  Drive-by reviews can find code nits, style, naming, etc without deep knowledge of an area, and allow people to participate without being granted any rights.

File: contributors/devel/owners.md (hunk 6 above)
Comment (spiffxp) [inline contributors/devel/owners.md]: Growing the pool of reviewers who aren't approvers would avoid this, but I view it is a symptom not a problem per se.  For me the main quirk is the slightly surprising behavior of a command being interpreted in one or two ways depending on who you are.

Comment (thockin) [inline contributors/devel/owners.md]: I always differentiate /lgtm and /approve, and often approve a PR, but let someone else do the review and lgtm.  I don't think there should be any overloading, and it's fine for one person to fill both roles, but explicitly so.

File: contributors/devel/owners.md (hunk 15 above)
Comment (spiffxp) [inline contributors/devel/owners.md]: Yeah still struggling here, I'll reword.  Anyone can review, and many eyes are encouraged.  Only members can `/lgtm`, and the anti-pattern is someone `/lgtm`'ing when they shouldn't.

File: contributors/devel/owners.md [hunk 16]
```diff
@@ -1,100 +1,240 @@
...
+### [`prow`](https://github.com/kubernetes/test-infra/tree/master/prow)
+
+Prow receives events from GitHub, and reacts to them. It is effectively stateless.
+
+- [plugin: lgtm](https://github.com/kubernetes/test-infra/tree/master/prow/plugins/lgtm)
+  - responsible for adding the `lgtm` label when a **reviewer** comments `/lgtm` on a PR
+  - the **PR author** may not `/lgtm` their own PR
+- [plugin: assign](https://github.com/kubernetes/test-infra/tree/master/prow/plugins/assign)
+  - responsible for assigning GitHub users in response to `/assign` comments on a PR
+  - responsible for unassigning GitHub users in response to `/unsassign` comments on a PR
```
Comment (cblecker) [inline contributors/devel/owners.md]: s/unsassign/unassign/

File: contributors/devel/owners.md (hunk 15 above)
Comment (thockin) [inline contributors/devel/owners.md]: well, the both should not apply an lgtm label except when triggered by an actual assignee, right?  So no REAL harm except confusion

Comment (cblecker) [inline contributors/devel/owners.md]: That's not the current state of the world today. If someone comments `/lgtm` on a PR, prow will attempt to assign the PR to them. If they are a collaborator or org member that is able to have the PR assigned to them, then it will do that, and then apply the LGTM label.

If you're suggesting that this be changed @thockin, then we should get an issue in test-infra to do so.

File: contributors/devel/owners.md (hunk 16 above)
Comment (spiffxp) [inline contributors/devel/owners.md]: nice catch

**Comment by spiffxp** [comment]
thanks all for the feedback, squashed

File: contributors/devel/owners.md (hunk 15 above)
Comment (thockin) [inline contributors/devel/owners.md]: I lost track of this, but I guess that IS what I was suggesting.
